File: Solutions/_15.go
package solutions

import "sort"

func main() {

}

func threeSum(nums []int) [][]int {
	res := [][]int{}
	sort.Ints(nums)

	for i, v := range nums {
		if i > 0 && v == nums[i-1] {
			continue
		}
		l := i + 1
		r := len(nums) - 1
		for l < r {
			sum := v + nums[l] + nums[r]
			if sum > 0 {
				r -= 1
			} else if sum < 0 {
				l += 1
			} else {
				res = append(res, []int{v, nums[l], nums[r]})
				l += 1
				for l < r && nums[l] == nums[l-1] {
					l += 1
				}
			}
		}
	}
	return res
}

File: Solutions/_15.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> threeSum(vector<int> &nums)
{
    sort(nums.begin(), nums.end());
    vector<vector<int>> result;

    for (int i = 0; i < nums.size(); i++)
    {
        if (i > 0 && nums[i] == nums[i - 1])
        {
            continue;
        }
        int left = i + 1;
        int right = nums.size() - 1;

        while (left < right)
        {
            int threeSum = nums[i] + nums[left] + nums[right];
            if (threeSum > 0)
            {
                right--;
            }
            else if (threeSum < 0)
            {
                left++;
            }
            else
            {
                result.push_back({nums[i], nums[left], nums[right]});
                left++;
                while (left < right && nums[left] == nums[left - 1])
                {
                    left++;
                }
            }
        }
    }

    return result;
}

int main()
{
    vector<int> nums = {-1, 0, 1, 2, -1, -4};
    vector<vector<int>> result = threeSum(nums);
    for (auto vec : result)
    {
        for (int val : vec)
        {
            cout << val << " ";
        }
        cout << endl;
    }
    return 0;
}